### PR TITLE
Add OpenThread Support to ESP32 DeviceLayer using ot-esp32

### PR DESCRIPTION
**Pull request by erjiaqing**
This PR is tested with https://github.com/openthread/ot-esp32 for thread support on ESP32.
Since ot-esp32 requires ESP-IDF 3.3 or later, this PR depends on https://github.com/openweave/openweave-core/pull/567. So only the last 2 commits need to be reviewed.

* ThreadStackManagerImpl is copied from EFR32 with some changes for
ot-esp32.
* ConnectivityManagerImpl and Warm is updated to support Thread if Enabled
* Removed EFR32_OPENTHREAD_API check in GenericThreadStackManagerImpl_OpenThread

### CODE REVIEW

**Comment by erjiaqing** [comment]
@robszewczyk PTAL, this PR based on #567, but I think it is also OK to ignore #567 and merge this directly.

**Comment by erjiaqing** [comment]
@robszewczyk I have rebased this PR, ptal.

File: src/adaptations/device-layer/Makefile.am [hunk 1]
```diff
@@ -162,13 +163,15 @@ libDeviceLayer_a_SOURCES               +=       \
...
     ESP32/ESP32Utils.cpp                        \
+    ESP32/ThreadStackManagerImpl.cpp            \ 
     ESP32/GroupKeyStoreImpl.cpp                 \
     ESP32/Logging.cpp                           \
     ESP32/LwIPCoreLock.cpp                      \
     ESP32/NetworkProvisioningServerImpl.cpp     \
     ESP32/PlatformManagerImpl.cpp               \
     ESP32/SoftwareUpdateManagerImpl.cpp         \
     ESP32/SystemTimeSupport.cpp                 \
+    OpenThread/OpenThreadUtils.cpp              \
```
Comment (bukepo) [inline src/adaptations/device-layer/Makefile.am:174]: keep sorted?

File: src/adaptations/device-layer/include/Weave/DeviceLayer/ESP32/WeavePlatformConfig.h [hunk 2]
```diff
@@ -110,4 +110,6 @@
 #define WEAVE_CONFIG_DATA_MANAGEMENT_ENABLE_SCHEMA_CHECK CONFIG_WDM_ENABLE_SCHEMA_CHECK
 #define WEAVE_CONFIG_DEFAULT_INCOMING_CONNECTION_IDLE_TIMEOUT CONFIG_DEFAULT_INCOMING_CONNECTION_IDLE_TIMEOUT
 
+#define WEAVE_PLATFORM_ESP32 1
```
Comment (bukepo) [inline src/adaptations/device-layer/include/Weave/DeviceLayer/ESP32/WeavePlatformConfig.h:113]:
```suggestion
#define WEAVE_PLATFORM_ESP32 1
```

File: src/adaptations/device-layer/include/Weave/DeviceLayer/OpenThread/GenericThreadStackManagerImpl_OpenThread.ipp [hunk 3]
```diff
@@ -1020,9 +1020,23 @@ WEAVE_ERROR GenericThreadStackManagerImpl_OpenThread<ImplClass>::DoInit(otInstan
     // create or acquire a singleton instance of OpenThread.
     if (otInst == NULL)
     {
+#ifdef WEAVE_PLATFORM_ESP32
```
Comment (bukepo) [inline src/adaptations/device-layer/include/Weave/DeviceLayer/OpenThread/GenericThreadStackManagerImpl_OpenThread.ipp:1023]:
```suggestion
#if defined(WEAVE_DEVICE_LAYER_TARGET_ESP32) && WEAVE_DEVICE_LAYER_TARGET_ESP32
```

File: src/adaptations/device-layer/ESP32/ThreadStackManagerImpl.cpp [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    self->mThreadTask = xTaskGetCurrentTaskHandle();
+
+    ThreadStackMgr().LockThreadStack();
+    otInstance *instance = ThreadStackMgrImpl().OTInstance();
+    ThreadStackMgr().UnlockThreadStack();
+    
+
+    while (true)
+    {
+      otSysMainloopContext mainloop;
```
Comment (lanyuwen) [inline src/adaptations/device-layer/ESP32/ThreadStackManagerImpl.cpp:70]: indents not aligned?

**Pull request merged**